src/main/java/fi/aalto/cs/intellij/activities/VersionCheckerActivity.java

```diff
@@ -0,0 +1,41 @@
+package fi.aalto.cs.intellij.activities;
+
+import com.intellij.notification.Notification;
+import com.intellij.notification.Notifications;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.startup.StartupActivity;
+import fi.aalto.cs.intellij.common.BuildInfo;
+import fi.aalto.cs.intellij.common.Version;
+import fi.aalto.cs.intellij.notifications.BetaVersionWarning;
+import fi.aalto.cs.intellij.notifications.Notifier;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * An activity that checks, on startup, if the version of the plugin is a pre-release, and, if that
+ * is the case, shows a {@link BetaVersionWarning}.
+ */
+public class VersionCheckerActivity implements StartupActivity {
+
+  private final Notifier notifier;
```

Comment:
Why do we need this instead of just using `Notifications.Bus::notify`?

Reply:
Just trying to decouple the VersionCheckerActivity from those parts of the framework that would make unit testing difficult. I mean, we could probably use Mockito to do some dependency injection to `Notifications.Bus.notify(...)` and the unit test would be like
```
public void testVersionChecker() {
  Mockito
    .doSomeMagicTo(Notifications.Bus.class)
    .soThatWhenSomeOneCalls('notify')
    .doNotDoWhatItWouldNormallyDoButInsteadDo((notification, project) -> {
      // Assert here that arguments are what they should be.
    })
    .andVerifyThatThisIsCalledExactlyThisManyTimes(1);

    new VersionCheckerActivity(new Version(...)).runActivity(...);
}
``` 
This seems actually OK to me. What I don't like a lot in this code is that the writer of the code has to "just know" that `VersionCheckerActivity.runActivity(...)` calls exactly that method (`Notifications.Bus.notify(...)`) of the framework.

I'd say that accessing `Notifications.Bus.notify(...)` through an interface makes it more explicit (for the client code) in which ways `VersionCheckerActivity` communicates with the framework and decouples it more from the implementation details of the framework.

Reply:
My understanding is that `Notifications.Bus.notify` is the "standard" way for plugins to display notifications, so in that sense I would actually prefer to assume that specifically `Notifications.Bus::notify` is called and get a failed test if it isn't called. Then again I'm not sure how difficult your pseudocode test approach would be to implement. Is there a (apart from testing) use case that uses a different notifier?

Reply:
Now thinking about it, it's quite convenient to be able to provide the notifier, but I still fail to see a "real world" use case

Reply:
" how difficult your pseudocode test approach would be to implement"

I've understood that Mockito is specifically designed to so things like that so it shouldn't be difficult. I think @superseacat has done some research on this.

Reply:
> I can't imagine a "real world" use case

The most important 'use case' for this is certainly unit testing, and I wouldn't underrate its importance.

Based on my experience, this is a very customary way to write code, and I have understood that making code unit-testable this way is considered a good practice (even though there are other approaches, too, I guess). I did some google search about how to write testable code and at least the first article in the results seems to highly recommend using this pattern so I don't think it's outdated by any means. Link to that article: 
https://www.toptal.com/qa/how-to-write-testable-code-and-why-it-matters

Reply:
Good point on the importance of testing; I'm convinced of this approach now :+1: